fix(plugin-content-docs): fix error message context (error cause) when doc processing fails

shanpriyan opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.
You can learn more about contributing to Docusaurus here: https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md
Happy contributing!
-->

## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [x] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

<!--
Please also remember to sign the CLA, although you can also sign it after submitting the PR. The CLA is required for us to merge your PR.
If this PR adds or changes functionality, please take some time to update the docs. You can also write docs after the API design is finalized and the code changes have been approved.
-->

## Motivation

<!-- Help us understand your motivation by explaining why you decided to make this change. Does this fix a bug? Does it close an issue? -->
`await` the promise from `doProcessDocMetadata` and throw error if it fails


<!-- Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos! -->

### Test links

<!--
🙏 Please add an exhaustive list of links relevant to this pull request.
⏱ This saves maintainers a lot of time during reviews.

- If you changed anything that's displayed on UI, please add a dogfooding page in website/_dogfooding to help us preview the effect. Those tests are deployed at https://docusaurus.io/tests
- If you changed documentation, please link to the new and updated documentation pages.

After submission, our Netlify bot will post a deploy preview link in comment, in the format of https://deploy-preview-<PR-NUMBER>--docusaurus-2.netlify.app/. Once available, please edit this section with links to the relevant deploy preview pages.

Please don't be afraid to change the main site's configuration as well! You can make use of your new feature on our site so we can preview its effects. We can decide if it should be kept in production before merging it.
-->

Deploy preview: https://deploy-preview-8234--docusaurus-2.netlify.app/

## Related issues/PRs
Resolves #8230
Suggested fix [#8230 (comment)](https://github.com/facebook/docusaurus/issues/8230#issuecomment-1285163941)

<!-- If you haven't already, link to issues/PRs that are related to this change. This helps us develop the context and keep a rich repo history. If this PR is a continuation of a past PR's work, link to that PR. If the PR addresses part of the problem in a meta-issue, mention that issue. -->

slorber commented on packages/docusaurus-plugin-content-docs/src/docs.ts:
```diff
@@ -316,18 +316,18 @@ async function doProcessDocMetadata({
 }): Promise<DocMetadataBase> {
   try {
-    return doProcessDocMetadata(args);
+    const docMetaData = await doProcessDocMetadata(args);
+    return docMetaData;
```
no need for intermediate var, just return await should be enough

slorber commented on packages/docusaurus-plugin-content-docs/src/docs.ts:
```diff
@@ -316,18 +316,18 @@ async function doProcessDocMetadata({
   } catch (err) {
-    logger.error`Can't process doc metadata for doc at path path=${args.docFile.filePath} in version name=${args.versionMetadata.versionName}`;
-    throw err;
+    throw new Error(`Can't process doc metadata for doc at path path=${args.docFile.filePath} in version name=${args.versionMetadata.versionName}`);
```
You can use Error Cause here:

https://2ality.com/2021/06/error-cause.html

slorber reviewed (changes requested):
good start thanks 👍 

Can you also please add a unit test somewhere so that we be sure it works and never stops working again?

shanpriyan commented:
>Can you also please add a unit test somewhere so that we be sure it works and never stops working again?

Actually I'm not very familiar with tests 😅, If you could guide me I'll add it. Any links similar tests would be appreciated

slorber commented:
You can get inspiration from: https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-plugin-content-docs/src/__tests__/docs.test.ts#L684

Create a md file with the bad frontmatter that should get rejected, then run the fn in a test and assert that the correct error message was thrown.

Run `yarn test docs.test.ts --watch` to run this specific test file. Jest provides `it.only(` and `describe.only(` to focus on a specific test case. 

Actually I think some existing tests might cover the error message already, and will start failing with your fix because the message is now updated to something better: you may have to change the assertions/snapshots to reflect the change. 

I'm not sure how Jest plays well with Error Cause, may have to look into it myself later. Let me know if you don't find a solution. If you want to contribute to open-source it will be important to be able to write tests so it's worth you give it a try ;)

slorber commented on packages/docusaurus-plugin-content-docs/src/docs.ts:
```diff
@@ -316,18 +316,17 @@ async function doProcessDocMetadata({
   } catch (err) {
-    logger.error`Can't process doc metadata for doc at path path=${args.docFile.filePath} in version name=${args.versionMetadata.versionName}`;
-    throw err;
+    throw new Error(`Can't process doc metadata for doc at path path=${args.docFile.filePath} in version name=${args.versionMetadata.versionName}`, {cause: err});
```
I think TS won't be happy because a catched error is of type "unknown"

you may need `err as Error`

shanpriyan commented:
>You can get inspiration from: https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-plugin-content-docs/src/__tests__/docs.test.ts#L684
>
>Create a md file with the bad frontmatter that should get rejected, then run the fn in a test and assert that the correct error message was thrown.
>
>Run yarn test docs.test.ts --watch to run this specific test file. Jest provides it.only( and describe.only( to focus on a specific test case.

This would be helpful to get started, thanks 

>If you want to contribute to open-source it will be important to be able to write tests so it's worth you give it a try ;)

😅 Will definitely consider your advice and I'll give this a try 🚀

shanpriyan commented:
>Actually I think some existing tests might cover the error message already, and will start failing with your fix because the message is now updated to something better: you may have to change the assertions/snapshots to reflect the change.

fixed existing test case (updated snapshot) which failed due to change in error message

|Before Fix|After fix|
|----|----|
|![image](https://user-images.githubusercontent.com/42532987/197329533-f26d59c9-324c-4b36-b31a-470a09370469.png)|![image](https://user-images.githubusercontent.com/42532987/197329553-c4e770ae-07f0-4263-8176-0afdf0a47877.png)

shanpriyan commented:
@slorber since the below _test case_ covers the latest error message (a532ed0576123a85b9fec6d354606009b5fa119d), Do I need to add a new _test case_? 

https://github.com/facebook/docusaurus/blob/e41133245ca4e6cee390da74ccfe305ccb2ddb25/packages/docusaurus-plugin-content-docs/src/__tests__/docs.test.ts#L684

slorber reviewed (approved):
Thanks 👍 

Fixed the test to assert that we have the full error cause chain being tested
